### PR TITLE
MVCC: Compact outdated data

### PR DESCRIPTION
**Pull request by c4pt0r**

### CODE REVIEW

File: kv/compactor.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kv
+
+import "time"
+
+type CompactorPolicy struct {
+	SafeTime        int
```
Comment (ngaut) [inline kv/compactor.go]: s/SafeTime/SafePoint

File: kv/kv.go [hunk 2]
```diff
@@ -164,6 +164,7 @@ type Storage interface {
 	Close() error
 	// Storage's unique ID
 	UUID() string
+	DumpRaw()
```
Comment (ngaut) [inline kv/kv.go]: Need comments.

File: store/localstore/compactor.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+var localCompactorDefaultPolicy = kv.CompactorPolicy{
+	SafeTime:        20 * 1000, // in ms
+	TriggerInterval: 1 * time.Second,
+	BatchDeleteSize: 100,
+}
+
+type localstoreCompactor struct {
+	mu         sync.Mutex
+	recentKeys map[string]struct{}
+	stopChan   chan struct{}
```
Comment (ngaut) [inline store/localstore/compactor.go]: s/Chan/Ch

File: store/localstore/compactor.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+	gc.mu.Lock()
+	defer gc.mu.Unlock()
+	gc.recentKeys[string(k)] = struct{}{}
+}
+
+func (gc *localstoreCompactor) getAllVersions(k kv.Key) ([]kv.EncodedKey, error) {
+	startKey := MvccEncodeVersionKey(k, kv.MaxVersion)
+	endKey := MvccEncodeVersionKey(k, kv.MinVersion)
+
+	it, err := gc.db.Seek(startKey)
```
Comment (ngaut) [inline store/localstore/compactor.go]: Duplicate code, could you refactor a function ScanRange(Range{startKey, endKey}, func)

File: store/localstore/compactor.go [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+		case <-gc.stopChan:
+			break L
+		case <-gc.ticker.C:
+			log.Info("GC trigger")
+			gc.mu.Lock()
+			m := gc.recentKeys
+			gc.recentKeys = make(map[string]struct{})
+			gc.mu.Unlock()
+			// Do Compactor
+			if len(m) > 0 {
```
Comment (ngaut) [inline store/localstore/compactor.go]: No need to check length of m

File: store/localstore/compactor.go [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+		_, ver, err := MvccDecode(k)
+		if err != nil {
+			// Should not happen.
+			panic(err)
+		}
+		ts := LocalVersionToTimestamp(ver)
+		currentTs := time.Now().UnixNano() / int64(time.Millisecond)
+		// Check timeout keys.
+		if currentTs-int64(ts) >= int64(gc.policy.SafeTime) {
+			// Skip first version.
```
Comment (ngaut) [inline store/localstore/compactor.go]: Change comments to "Keep the most recent version which close to safe point"

File: store/localstore/compactor.go [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+		log.Info("GC send key to deleteWorker", key)
+		gc.delChan <- key
+	}
+	return nil
+}
+
+func (gc *localstoreCompactor) Start() {
+	// Start workers.
+	go gc.checkExpiredKeysWorker()
+	for i := 0; i < deleteWorkerSize; i++ {
```
Comment (ngaut) [inline store/localstore/compactor.go]: s/deleteWorkerSize/deleteWorkerCnt/g

File: store/localstore/compactor.go [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+func (gc *localstoreCompactor) Stop() {
+	gc.ticker.Stop()
+	close(gc.stopChan)
+}
+
+func newLocalCompactor(policy kv.CompactorPolicy, db engine.DB) *localstoreCompactor {
+	return &localstoreCompactor{
+		recentKeys: make(map[string]struct{}),
+		stopChan:   make(chan struct{}),
+		delChan:    make(chan kv.EncodedKey, 100),
```
Comment (ngaut) [inline store/localstore/compactor.go]: define 100 as constant.

File: kv/compactor.go [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+// CompactorPolicy defines gc policy of MVCC storage.
+type CompactorPolicy struct {
+	// SafePoint specifies
+	SafePoint int
+	// TriggerInterval specifies how often should the compactor
+	// scans outdated data.
+	TriggerInterval time.Duration
+	// BatchDeleteCnt specifies the batch size for
+	// delete outdated data transaction.
```
Comment (shenli) [inline kv/compactor.go]: deleting

File: store/localstore/compactor.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+		}
+	}
+}
+
+func (gc *localstoreCompactor) checkExpiredKeysWorker() {
+L:
+	for {
+		select {
+		case <-gc.stopCh:
+			break L
```
Comment (shenli) [inline store/localstore/compactor.go]: Why not use return?

File: kv/compactor.go [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kv
+
+import "time"
+
+// CompactorPolicy defines gc policy of MVCC storage.
+type CompactorPolicy struct {
+	// SafePoint specifies
```
Comment (shenli) [inline kv/compactor.go]: The comment for SafePoint is not clear.

File: kv/compactor.go [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+	SafePoint int
```
Comment (qiuyesuifeng) [inline kv/compactor.go]: How about SafeMark?

Comment (qiuyesuifeng) [inline kv/compactor.go]: Comment end of .

File: kv/compactor.go (hunk 9 above)
Comment (qiuyesuifeng) [inline kv/compactor.go]: per transaction or in transaction?

File: kv/compactor.go [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+	// BatchDeleteCnt specifies the batch size for
+	// delete outdated data transaction.
+	BatchDeleteCnt int
+}
+
+// Compactor compacts MVCC storage.
+type Compactor interface {
+	// OnGet is the hook point on Txn.Get.
+	OnGet(k Key)
+	// OnSet is the hook point on Txn.Set
```
Comment (qiuyesuifeng) [inline kv/compactor.go]: Comment end of .

File: store/localstore/compactor.go (hunk 10 above)
Comment (qiuyesuifeng) [inline store/localstore/compactor.go]: +1

File: kv/compactor.go (hunk 12 above)
Comment (shenli) [inline kv/compactor.go]: ShelfLife is a better name. This is a time slot, not a time point.

File: store/localstore/compactor.go (hunk 4 above)
Comment (c4pt0r) [inline store/localstore/compactor.go]: refactor these code involved redesign dbSnapshot's buffered iterator.  I'll open another PR to do this.

**Comment by c4pt0r** [comment]
@shenli @qiuyesuifeng @ngaut  PTAL

File: kv/compactor.go [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+	BatchDeleteCnt int
+}
+
+// Compactor compacts MVCC storage.
+type Compactor interface {
+	// OnGet is the hook point on Txn.Get.
+	OnGet(k Key)
+	// OnSet is the hook point on Txn.Set.
+	OnSet(k Key)
+	// OnDelete is the hook point on Txn.Delete
```
Comment (qiuyesuifeng) [inline kv/compactor.go]: Comment end of .

File: store/localstore/compactor_test.go [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+type localstoreCompactorTestSuite struct {
+}
+
+func count(db engine.DB) int {
+	it, _ := db.Seek([]byte{0})
+	defer it.Release()
+	totalCnt := 0
+	for it.Next() {
+		log.Error(it.Key())
```
Comment (qiuyesuifeng) [inline store/localstore/compactor_test.go]: Remove this log?

File: kv/compactor.go (hunk 14 above)
Comment (c4pt0r) [inline kv/compactor.go]: DONE

File: store/localstore/compactor_test.go [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+	policy := kv.CompactPolicy{
+		SafePoint:       500,
+		BatchDeleteCnt:  1,
+		TriggerInterval: 100 * time.Millisecond,
+	}
+	compactor := newLocalCompactor(policy, db)
+	store.(*dbStore).compactor = compactor
+
+	compactor.Start()
```
Comment (qiuyesuifeng) [inline store/localstore/compactor_test.go]: Need  compactor.Stop()?

Comment (c4pt0r) [inline store/localstore/compactor_test.go]: DONE

File: store/localstore/compactor.go [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+	first := true
+	// keys are always in descending order.
+	for _, k := range keys {
+		_, ver, err := MvccDecode(k)
+		if err != nil {
+			// Should not happen.
+			panic(err)
+		}
+		ts := localVersionToTimestamp(ver)
+		currentTs := time.Now().UnixNano() / int64(time.Millisecond)
```
Comment (ngaut) [inline store/localstore/compactor.go]: s/Ts/TS

**Comment by ngaut** [comment]
LGTM

**Pull request merged**